# ***CLEANING PROCESS***

In [3]:
import pandas as pd
import matplotlib as plt

In [4]:
booking=pd.read_csv("/content/fact_bookings.csv")
agg_booking=pd.read_csv("/content/fact_aggregated_bookings.csv")


In [ ]:
booking.shape

(134590, 12)

"In this 'no_guests' column, there are negative values, which are not valid, so we will ignore them."

In [9]:
clean_booking=booking[booking.no_guests>0]
clean_booking.head(5)

booking_id  property_id booking_date check_in_date checkout_date  \
1  May012216558RT12      16558.0     30-04-22      1/5/2022      2/5/2022   
2  May012216558RT13      16558.0     28-04-22      1/5/2022      4/5/2022   
4  May012216558RT15      16558.0     27-04-22      1/5/2022      2/5/2022   
5  May012216558RT16      16558.0     1/5/2022      1/5/2022      3/5/2022   
6  May012216558RT17      16558.0     28-04-22      1/5/2022      6/5/2022   

   no_guests room_category booking_platform  ratings_given booking_status  \
1        2.0           RT1           others            NaN      Cancelled   
2        2.0           RT1          logtrip            5.0    Checked Out   
4        4.0           RT1    direct online            5.0    Checked Out   
5        2.0           RT1           others            4.0    Checked Out   
6        2.0           RT1           others            NaN      Cancelled   

   revenue_generated  revenue_realized  
1             9100.0            3640.0  
2          9100000.0            9100.0  
4            10920.0           10920.0  
5             9100.0            9100.0  
6             9100.0            3640.0

"Now as you can see the difference in shape value"

In [ ]:
clean_booking.shape

(91760, 12)

"Now we have to find out the outliers"

In [ ]:
min,max=booking.revenue_generated.min(),booking.revenue_generated.max()
min,max

(6500.0, 28560000.0)

In [ ]:
mean,std=booking.revenue_generated.mean(),booking.revenue_generated.median()
mean,std

(15378.05412734973, 13500.0)

In [ ]:
higher_limit=mean + 3 * std
higher_limit

55878.05412734973

In [ ]:
lower_limit=mean-3*std
lower_limit


-307456.64118583227

In [ ]:
booking['revenue_generated'].describe()

count    9.177000e+04
mean     1.549507e+04
std      1.076506e+05
min      6.500000e+03
25%      9.900000e+03
50%      1.350000e+04
75%      1.800000e+04
max      2.856000e+07
Name: revenue_generated, dtype: float64

In [ ]:
booking[booking.revenue_generated>higher_limit]

booking_id  property_id booking_date check_in_date checkout_date  \
2     May012216558RT13        16558     28-04-22      1/5/2022      4/5/2022   
111   May012216559RT32        16559     29-04-22      1/5/2022      2/5/2022   
315   May012216562RT22        16562     28-04-22      1/5/2022      4/5/2022   
562  May012217559RT118        17559     26-04-22      1/5/2022      2/5/2022   

     no_guests room_category booking_platform  ratings_given booking_status  \
2          2.0           RT1          logtrip            5.0    Checked Out   
111        6.0           RT3    direct online            NaN    Checked Out   
315        2.0           RT2   direct offline            3.0    Checked Out   
562        2.0           RT1           others            NaN      Cancelled   

     revenue_generated  revenue_realized  
2            9100000.0            9100.0  
111         28560000.0           28560.0  
315         12600000.0           12600.0  
562          2000000.0            4420.0

In [ ]:
bookings=booking[booking.revenue_generated<=higher_limit]
bookings.shape

(91766, 12)

In [ ]:
min,max=booking.revenue_realized.min(),booking.revenue_realized.max()
min,max

(2600, 45220)

In [ ]:
higher_limit = booking.revenue_realized.mean() + 3*booking.revenue_realized.std()
higher_limit

33480.44762788345

In [ ]:
booking[booking.revenue_realized>higher_limit].head(5)


booking_id  property_id booking_date check_in_date checkout_date  \
137   May012216559RT41        16559     27-04-22      1/5/2022      7/5/2022   
139   May012216559RT43        16559     1/5/2022      1/5/2022      2/5/2022   
143   May012216559RT47        16559     28-04-22      1/5/2022      3/5/2022   
149  May012216559RT413        16559     24-04-22      1/5/2022      7/5/2022   
222   May012216560RT45        16560     30-04-22      1/5/2022      3/5/2022   

     no_guests room_category booking_platform  ratings_given booking_status  \
137        4.0           RT4           others            NaN    Checked Out   
139        6.0           RT4         tripster            3.0    Checked Out   
143        3.0           RT4           others            5.0    Checked Out   
149        5.0           RT4          logtrip            NaN    Checked Out   
222        5.0           RT4           others            3.0    Checked Out   

     revenue_generated  revenue_realized  
137              38760             38760  
139              45220             45220  
143              35530             35530  
149              41990             41990  
222              34580             34580

"One observation we can have in above dataframe is that all rooms are RT4 which means presidential suit. Now since RT4 is a luxurious room it is likely their rent will be higher. To make a fair analysis, we need to do data analysis only on RT4 room types"

In [ ]:
booking[booking.room_category=="RT4"].revenue_realized.describe()

count    16073.000000
mean     23440.103652
std       9048.865206
min       7600.000000
25%      19000.000000
50%      26600.000000
75%      32300.000000
max      45220.000000
Name: revenue_realized, dtype: float64

In [ ]:
# mean + 3*standard deviation
higher_limit=23440+3*9048
higher_limit

50584

**The calculated higher limit is 50,583, and in the above DataFrame, the maximum value for revenue realized is 45,220. Therefore, we can conclude that there are no outliers in this column, and no data cleaning is required for this particular field.**

In [ ]:
booking[booking.booking_id=="May012216558RT213"]

booking_id  property_id booking_date check_in_date checkout_date  \
30  May012216558RT213        16558     29-04-22      1/5/2022      2/5/2022   

    no_guests room_category booking_platform  ratings_given booking_status  \
30        NaN           RT2          logtrip            4.0    Checked Out   

    revenue_generated  revenue_realized  
30              12600             12600

In [ ]:
booking.isnull().sum()

booking_id               0
property_id              0
booking_date             0
check_in_date            0
checkout_date            0
no_guests                3
room_category            0
booking_platform         0
ratings_given        77907
booking_status           0
revenue_generated        0
revenue_realized         0
dtype: int64

**The total number of values in our DataFrame is 134,576, out of which 77,899 rows have a null rating. Since a significant portion of the data has null ratings, we should neither filter out these rows nor replace the missing ratings with statistical measures like the mean or median.**

**In the aggregate bookings data, identify the columns containing null values. Replace these null values with an appropriate substitute, such as the mean or median, based on what you deem most suitable.**

In [ ]:
agg_booking.isnull().sum()

property_id            0
check_in_date          0
room_category          0
successful_bookings    0
capacity               2
dtype: int64

In [ ]:
agg_booking[agg_booking.capacity.isna()]

property_id check_in_date room_category  successful_bookings  capacity
8         17561      1-May-22           RT1                   22       NaN
14        17562      1-May-22           RT1                   12       NaN

In [ ]:
agg_booking.capacity.median()

25.0

In [ ]:
agg_booking.loc[[8,15]]

property_id check_in_date room_category  successful_bookings  capacity
8         17561      1-May-22           RT1                   22      25.0
15        17563      1-May-22           RT1                   21      25.0

**In the aggregate bookings data, identify the records where the successful_bookings value exceeds the capacity. Filter out these records for further analysis.**

In [ ]:
agg_booking[agg_booking.successful_bookings>agg_booking.capacity]

property_id check_in_date room_category  successful_bookings  capacity
3           17558      1-May-22           RT1                   30      19.0
12          16563      1-May-22           RT1                  100      41.0
4136        19558     11-Jun-22           RT2                   50      39.0
6209        19560      2-Jul-22           RT1                  123      26.0
8522        19559     25-Jul-22           RT1                   35      24.0
9194        18563     31-Jul-22           RT4                   20      18.0

In [ ]:
agg_booking.shape

(9200, 5)

In [ ]:
agg_booking = agg_booking[agg_booking.successful_bookings<=agg_booking.capacity]
agg_booking.shape

(9194, 5)

In [ ]:
agg_booking = agg_booking[agg_booking.successful_bookings<=agg_booking.capacity].head(5)
agg_booking

property_id check_in_date room_category  successful_bookings  capacity
0        16559      1-May-22           RT1                   25      30.0
1        19562      1-May-22           RT1                   28      30.0
2        19563      1-May-22           RT1                   23      30.0
4        16558      1-May-22           RT1                   18      19.0
5        17560      1-May-22           RT1                   28      40.0